### PR TITLE
fix: Fix manual compaction logic

### LINKED ISSUE #158: Bug: `DB::flush` miss
## Bug Report
during manual compaction, there is no compaction because there is no trigger threshold.

### PR DESCRIPTION
**Pull request by jonathanc-n**
Closes #158.

### Added self compaction on l0. 
 If we have manual compaction on level 0 and level 1 is empty. Due to level 0 SSTs being unsorted we can self compact level 0 SSTs that fall in the key range.

### Fixed the logic in `this_level_scopes`
The logic in `this_level_scopes` was incorrect as it would do a linear search and stop when a SST didnt fall in the range. However level 0 SSTs are unsorted and this would lead to incorrect behaviour. 

Instead I changed it so that it would do a linear search over all the SSTs and append every SST that fell in the range.

### Fixed the original issue 
Manual compaction will now be used even if threshold is not met.

### CODE REVIEW

**Comment by jonathanc-n** [comment]
I think we should also include a non-manual flush

File: src/compaction/leveled.rs [hunk 1]
```diff
@@ -987,4 +1088,372 @@ pub(crate) mod tests {
         }
         dbg!(version);
     }
+
+    // Self compaction is when on level 0 there is no files in the next level
+    #[tokio::test(flavor = "multi_thread")]
+    async fn test_manual_self_compaction() {
```
Comment (jonathanc-n) [inline src/compaction/leveled.rs:1094]: We should have many more test like these to test compaction edge cases.

**Review by ethe — APPROVED** [review]

File: src/compaction/leveled.rs [hunk 2]
```diff
@@ -204,14 +205,87 @@ where
...
+                                    ProjectionMask::all(),
+                                )
+                                .await?,
+                        });
+                    }
+                    Compactor::<R>::build_tables(
+                        option,
+                        version_edits,
+                        level,
+                        streams,
```
Comment (crwen) [inline src/compaction/leveled.rs:265]: Can we push to level + 1 ?

**Comment by jonathanc-n** [comment]
@crwen This is fixed, good for another review!
